diff --git a/src/Illuminate/Database/Connection.php b/src/Illuminate/Database/Connection.php
index 7021a40..ae9cb0c 100755
--- a/src/Illuminate/Database/Connection.php
+++ b/src/Illuminate/Database/Connection.php
@@ -486,14 +486,22 @@
 	 */
 	public function beginTransaction()
 	{
-		++$this->transactions;
+            if($this->transactions === 0) {
+                try {
+                    $this->pdo->beginTransaction();
+                } catch (\Exception $e) {
+                    if($this->causedByLostConnection($e)) {
+                        $this->reconnect();
+                        $this->pdo->beginTransaction();
+                    } else {
+                        throw $e;
+                    }
+                }
+            }
+            
+            ++$this->transactions;
 
-		if ($this->transactions == 1)
-		{
-			$this->pdo->beginTransaction();
-		}
-
-		$this->fireConnectionEvent('beganTransaction');
+            $this->fireConnectionEvent('beganTransaction');
 	}
 
 	/**
@@ -659,18 +667,31 @@
 		throw $e;
 	}
 
-	/**
-	 * Determine if the given exception was caused by a lost connection.
-	 *
-	 * @param  \Illuminate\Database\QueryException
-	 * @return bool
-	 */
-	protected function causedByLostConnection(QueryException $e)
-	{
-		return str_contains($e->getPrevious()->getMessage(), 'server has gone away');
-	}
+        /**
+         * Determine if the given exception was caused by a lost connection.
+         *
+         * @param  \Exception
+         * @return bool
+         */
+        protected function causedByLostConnection(\Exception $e)
+        {
+            $message = $e->getMessage();
 
-	/**
+            return \Illuminate\Support\Str::contains($message, [
+                        'server has gone away',
+                        'no connection to the server',
+                        'Lost connection',
+                        'is dead or not enabled',
+                        'Error while sending',
+                        'decryption failed or bad record mac',
+                        'server closed the connection unexpectedly',
+                        'SSL connection has been closed unexpectedly',
+                        'Error writing data to the connection',
+                        'Resource deadlock avoided',
+            ]);
+        }
+
+        /**
 	 * Disconnect from the underlying PDO connection.
 	 *
 	 * @return void

